### ISSUE TITLE
Nintendo Switch: LuaJIT interpreter for ARM64

### ISSUE THREAD

**Issue by jpanettiere** [LuaJIT/LuaJIT#347]
I'm seeing a bug when I try running LuaJIT interpreter-mode for ARM64 (using LuaJIT-2.1.0-beta3).

In C++, the following code is executed:
`		lua_newtable(L);

		if (argc > 0)
		{
			lua_pushstring(L, argv[0]);
			lua_rawseti(L, -2, -2);
		}

		lua_pushstring(L, "embedded boot.lua");
		lua_rawseti(L, -2, -1);

		for (int i = 1; i < argc; i++)
		{
			lua_pushstring(L, argv[i]);
			lua_rawseti(L, -2, i);
		}

		lua_setglobal(L, "arg");
`

Then the following Lua code is issued:
`print('love.boot')
print('arg[-1]=', arg[-1])
print('arg[-2]=', arg[-2])
print('args provided:')
for k,v in pairs(arg) do
                print('k=', k, 'v=', v)
end
print('done checking args.')
`
The output should be as follows:

> love.boot
> arg[-1]=                embedded boot.lua
> arg[-2]=                Game
> args provided:
> k=           -2            v=           Game
> k=           -1            v=           embedded boot.lua
> done checking args.

But instead we're getting this:

> love.boot
> arg[-1]=      embedded boot.lua
> arg[-2]=      Game
> args provided:
> done checking args.

In other words, the pairs() iteration isn't fetching the table values.

I just wanted to mention this in case it wasn't already a known issue.  If this is the wrong place, please let me know where I should send this report.  I'd also be curious to know when the next build of LuaJIT for ARM64 can be expected.

Thank you!

John

**Comment by sindrom91** [comment]
Which OS are we talking about? iOS? Android? I cannot reproduce this on ARM64 Linux.

**Comment by jpanettiere** [comment]
We are porting this to Nintendo Switch.  Here is the batch file I made to build the library (copied from the PS4 batch file, but I swapped in ARM64 where relevant):

`@set LJCOMPILE=cl /nologo /c /MD /O2 /W3 /D_CRT_SECURE_NO_DEPRECATE
@set LJLINK=link /nologo
@set LJMT=mt /nologo
@set DASMDIR=..\dynasm
@set DASM=%DASMDIR%\dynasm.lua
@set ALL_LIB=lib_base.c lib_math.c lib_bit.c lib_string.c lib_table.c lib_io.c lib_os.c lib_package.c lib_debug.c lib_jit.c lib_ffi.c
@set GC64=-DLUAJIT_ENABLE_GC64
@set DASC=vm_arm64.dasc

@if "%1" neq "gc32" goto :NOGC32
@shift
@set GC64=
@set DASC=vm_arm.dasc
:NOGC32

%LJCOMPILE% host\minilua.c
@if errorlevel 1 goto :BAD
%LJLINK% /out:minilua.exe minilua.obj
@if errorlevel 1 goto :BAD
if exist minilua.exe.manifest^
  %LJMT% -manifest minilua.exe.manifest -outputresource:minilua.exe

@rem Check for 64 bit host compiler.
@minilua
@if not errorlevel 8 goto :FAIL

@set DASMFLAGS=-D P64 -D NO_UNWIND
minilua %DASM% -LN %DASMFLAGS% -o host\buildvm_arch.h %DASC%
@if errorlevel 1 goto :BAD

%LJCOMPILE% /I "." /I %DASMDIR% %GC64% -DLUAJIT_TARGET=LUAJIT_ARCH_ARM64 -DLUAJIT_OS=LUAJIT_OS_OTHER -DLUAJIT_DISABLE_JIT -DLUAJIT_DISABLE_FFI -DLUAJIT_NO_UNWIND host\buildvm*.c
@if errorlevel 1 goto :BAD
%LJLINK% /out:buildvm.exe buildvm*.obj
@if errorlevel 1 goto :BAD
if exist buildvm.exe.manifest^
  %LJMT% -manifest buildvm.exe.manifest -outputresource:buildvm.exe

buildvm -m elfasm -o lj_vm.s
@if errorlevel 1 goto :BAD
buildvm -m bcdef -o lj_bcdef.h %ALL_LIB%
@if errorlevel 1 goto :BAD
buildvm -m ffdef -o lj_ffdef.h %ALL_LIB%
@if errorlevel 1 goto :BAD
buildvm -m libdef -o lj_libdef.h %ALL_LIB%
@if errorlevel 1 goto :BAD
buildvm -m recdef -o lj_recdef.h %ALL_LIB%
@if errorlevel 1 goto :BAD
buildvm -m vmdef -o jit\vmdef.lua %ALL_LIB%
@if errorlevel 1 goto :BAD
buildvm -m folddef -o lj_folddef.h lj_opt_fold.c
@if errorlevel 1 goto :BAD

goto SkipOrbis

@rem ---- Cross compiler ----
@set LJCOMPILE="%SCE_ORBIS_SDK_DIR%\host_tools\bin\orbis-clang" -c -Wall -DLUAJIT_DISABLE_FFI %GC64%
@set LJLIB="%SCE_ORBIS_SDK_DIR%\host_tools\bin\orbis-ar" rcus
@set INCLUDE=""

orbis-as -o lj_vm.o lj_vm.s

@if "%1" neq "debug" goto :NODEBUG
@shift
@set LJCOMPILE=%LJCOMPILE% -g -O0
@set TARGETLIB=libluajitD_ps4.a
goto :BUILD
:NODEBUG
@set LJCOMPILE=%LJCOMPILE% -O2
@set TARGETLIB=libluajit_ps4.a
:BUILD
del %TARGETLIB%
@if "%1" neq "noamalg" goto :AMALG
for %%f in (lj_*.c lib_*.c) do (
  %LJCOMPILE% %%f
  @if errorlevel 1 goto :BAD
)

%LJLIB% %TARGETLIB% lj_*.o lib_*.o
@if errorlevel 1 goto :BAD
@goto :NOAMALG
:AMALG
%LJCOMPILE% ljamalg.c
@if errorlevel 1 goto :BAD
%LJLIB% %TARGETLIB% ljamalg.o lj_vm.o
@if errorlevel 1 goto :BAD
:NOAMALG

`
PS4 built the lib in the batch file, but I just used the above to make the source files and then added lib_*.c, lj_*.c and lj_vm.s to a static library in Visual Studio, using the following macro defines in the project:

LUAJIT_OS=LUAJIT_OS_OTHER;LUAJIT_TARGET=LUAJIT_ARCH_ARM64;LUAJIT_DISABLE_FFI;LUAJIT_DISABLE_JIT;

In addition to the default macros defined for Nintendo Switch.

Thank you!

**Comment by corsix** [comment]
Given `@set GC64=-DLUAJIT_ENABLE_GC64` in your batch file, should `LUAJIT_ENABLE_GC64` be a macro define in your project?

**Comment by jpanettiere** [comment]
That should say "lib_(asterisk).c" and "lj_(asterisk).c", apparently that caused some formatting thing in the text.

**Comment by jpanettiere** [comment]
Let me give that a try, thank you.

**Comment by corsix** [comment]
Possibly ditto `LUAJIT_NO_UNWIND`, though `GC64` is the major one.

**Comment by jpanettiere** [comment]
So I tried defining the macros as follows:

LUAJIT_NO_UNWIND;LUAJIT_ENABLE_GC64;LUAJIT_OS=LUAJIT_OS_OTHER;LUAJIT_TARGET=LUAJIT_ARCH_ARM64;LUAJIT_DISABLE_FFI;LUAJIT_DISABLE_JIT;

When I do that I get:

luajit.a(lj_vm.o):(.eh_frame+0x12): undefined reference to `lj_err_unwind_dwarf'

I tried removing the LUAJIT_NO_UNWIND macro but leaving the rest, and I'm still getting the same error, it doesn't step through the pairs.

**Comment by jpanettiere** [comment]
I just tried rebuilding the source using my batch file - which does define NO_UNWIND - and I confirmed that the resultant lj_vm.s contains the following code:

`.LSCIE1:
	.long 0
	.byte 0x1
	.string "zPR"
	.uleb128 0x1
	.sleb128 -8
	.byte 30
	.uleb128 6
	.byte 0x1b
	.long lj_err_unwind_dwarf-.
	.byte 0x1b
	.byte 0xc
	.uleb128 31
	.uleb128 0
	.align 3
`

**Comment by sindrom91** [comment]
It's not your error. Regular ARM64 Linux build fails when compiled with `-DLUAJIT_NO_UNWIND`.

**Comment by jpanettiere** [comment]
Thank you.  I tried rebuilding with LUAJIT_NO_UNWIND removed entirely, and that fixed the link error, but still getting the pairs() issue.

One thing I wanted to raise with you:  I'm using a slightly older version of the Nintendo SDK, in which several standard lib functions were not defined, including mmap.  I looked at the context in which it was called and it seemed to be just looking to allocate a memory block, so I defined my own mmap which returns malloc(length).  From there it goes immediately into:

`    uintptr_t addr = (uintptr_t)p;
    if ((addr >> LJ_ALLOC_MBITS) == 0 && addr >= LJ_ALLOC_MMAP_PROBE_LOWER) {
      /* We got a suitable address. Bump the hint address. */
      hint_addr = addr + size;
      errno = olderr;
      return p;
    }
`

in lj_alloc.c.  So it looks to me like it got what it wanted, but if you think this might be the problem, please let me know.

I'm going to discuss the possibility of updating our SDK with my dev team in case the issue lies in the older version of the SDK, of course there are potential problems with doing that as well.  If you want me to send you the source files created by my build script, or any other information, please let me know - we're eager to figure out what's going wrong so we can get Luajit support in our application.

Thank you!

**Comment by liangdong-xd** [comment]
@jpanettiere Do you solve this problem? I am going to try to compile luajit at NintendoSwitch. I don't konw how to do it.

**Comment by Hydroque** [comment]
liangdong-xd: Makefile --dry-run and go from there. It uses simplex gcc commands to build it as a so and dll. You should translate accordingly.

**Comment by cigumo** [comment]
@jpanettiere  I'm also interested in knowing if it worked. Thank you.

**Comment by jpanettiere** [comment]
So sorry it took me so long to reply here.  For our project, we ended up changing our Switch project to 32-bit and used the ARM32 version of LuaJIT.  I was not able to get the 64-bit version working.  When changing to 32-bit, we discovered that we had a much lower memory ceiling, so please bear that in mind if you decide to go 32-bit as well @liangdong-xd @cigumo .

**Comment by jpanettiere** [comment]
It wasn't hard, you select the NX32 config and everything else should just fall into place.  You might have to copy over some of your preprocessor defines, libraries, etc from the NX64 config.  Yea, I think you can only do interpreted mode on console.  We were a little worried that we might lose performance running in 32-bit mode, but the LuaJIT more than made up for it.

**Comment by cigumo** [comment]
That's great to know!
Thank you!

**Comment by bryanmcconkey** [comment]
I ran into the same problem and managed to find the resolution. Not sure what I can post here (publicly) so I posted my solution on the Nintendo forums if you still need a fix.

**Comment by IoriBranford** [comment]
How are you all avoiding this error:
```
src\lj_jit.h(442): error C2229: struct 'jit_State' has an illegal zero-sized array
src\lj_jit.h(445): error C2229: struct 'jit_State' has an illegal zero-sized array
```
The arrays in question are k32 and k64. Here I see their sizes defined as 0 for ARM64 (and for ARM as well so compiling 32-bit has the same problem).
```
enum {
#if LJ_TARGET_X86ORX64
  LJ_K64_TOBIT,		/* 2^52 + 2^51 */
  LJ_K64_2P64,		/* 2^64 */
  LJ_K64_M2P64,		/* -2^64 */
#if LJ_32
  LJ_K64_M2P64_31,	/* -2^64 or -2^31 */
#else
  LJ_K64_M2P64_31 = LJ_K64_M2P64,
#endif
#endif
#if LJ_TARGET_MIPS
  LJ_K64_2P31,		/* 2^31 */
#if LJ_64
  LJ_K64_2P63,		/* 2^63 */
  LJ_K64_M2P64,		/* -2^64 */
#endif
#endif
  LJ_K64__MAX,
};

enum {
#if LJ_TARGET_X86ORX64
  LJ_K32_M2P64_31,	/* -2^64 or -2^31 */
#endif
#if LJ_TARGET_PPC
  LJ_K32_2P52_2P31,	/* 2^52 + 2^31 */
  LJ_K32_2P52,		/* 2^52 */
#endif
#if LJ_TARGET_PPC || LJ_TARGET_MIPS
  LJ_K32_2P31,		/* 2^31 */
#endif
#if LJ_TARGET_MIPS64
  LJ_K32_2P63,		/* 2^63 */
  LJ_K32_M2P64,		/* -2^64 */
#endif
  LJ_K32__MAX
};
```

**Comment by MikePall** [comment]
Closing here, since nobody dares to submit a (probably trivial) patch.
